src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -129,17 +129,43 @@ public void close() {
             }
         }
     }
+    
+    private void ensureHasNext() {
+        if (!currentBlock.hasNext()) {
+            // Rethrow any exceptions raised on the background thread
+            // at the point the exception would have been encountered
+            // if we had performed synchronous iteration
+            raiseBackgroundThreadException();
```

Comment:
By throwing at the end of the block, the exception timing will match the exception timing from iterating over the underlying iterator. If your underlying iterator blows up on a particular record with a 100 record async buffer, you want your aysnc iterator to also blow up on that same record - not a record somewhere between 0 and 99 records beforehand.

By delaying the exception timing, the async iterator will have the same exception behavior as using the iterator directly.